Add Dynamic Scan support for partitioned tables using Orca

chrishajas opened the pull request:
Previously, we added support in ORCA for the postgres 10 style of
partitioning and utilized the postgres seq/index/bitmap executor nodes.
However, for plans with many partitions and many columns, this resulted
in large plan sizes (on the order of 100MB+). When combined with many
slices, this would cause increased memory utilization, sometimes going
OOM.

This series of commits reverts to using the GPDB6 style of Dynamic executor scan nodes.
We are keeping the logic to read from the GPDB7 catalog and the internal
Orca changes to handle static partition elimination within Orca. The
only intention of these changes is to use Dynamic executor nodes and
reduce the plan size. The plan shapes and performance are the same.

Some numbers: we found that for tables with 1k+ partitions and 1k+
columns, the dispatch plan size for a single `select * from foo` was
100MB vs 300KB with this new approach. By itself, 100MB may not seem
like a lot, but if there are multiple partitioned tables with multiple
joins and slices, this can quickly become quite a bit.

Many of the changes here were copied and adapted from GPDB6. However, we are performing
static partition elimination within the optimizer and therefore no longer need a 
partition selector when doing static elimination. We ran performance and workload tests,
and didn't see any unexpected plan changes.

Co-authored-by: Chris Hajas <chajas@vmware.com>
Co-authored-by: gpopt <103469259+gpopt@users.noreply.github.com>

gpopt commented on src/backend/executor/nodeDynamicIndexscan.c:
```diff
@@ -0,0 +1,343 @@
+{
+	DynamicIndexScanState *dynamicIndexScanState;
+    ListCell *lc;
+    int i;
```
The indentation looks odd.

gpopt commented on src/backend/executor/nodeDynamicIndexscan.c:
```diff
@@ -0,0 +1,343 @@
+	 * open the base relation and acquire appropriate lock on it.
+	 */
+	currentRelation = table_open(tableOid, AccessShareLock);
+    node->ss.ss_currentRelation = currentRelation;
```
indent

gpopt replied:
It is not fixed.

gpopt commented on src/backend/executor/nodeDynamicIndexscan.c:
```diff
@@ -0,0 +1,343 @@
+	{
+		/* This is the oid of a partition of the table (*not* index) */
+		Oid			*tableOid;
+        if (++node->whichPart < node->nOids)
```
indent

gpopt commented on src/backend/executor/nodeDynamicIndexscan.c:
```diff
@@ -0,0 +1,343 @@
+        foreach_with_count(lc, newPartOids, i)
+            node->partOids[i] = lfirst_oid(lc);
+        node->nOids = list_length(newPartOids);
+    }
```
The indent from L237~L266 is using spaces, not tabs.

gpopt commented on src/backend/executor/nodeDynamicIndexscan.c:
```diff
@@ -0,0 +1,343 @@
+	}
+	node->scan_state = SCAN_INIT;
+    // reset partition internal state
+    node->whichPart = -1;
```
indent

gpopt commented on src/backend/executor/nodeIndexscan.c:
```diff
@@ -902,190 +902,199 @@ ExecIndexRestrPos(IndexScanState *node)
-	indexstate->iss_RuntimeKeysReady = false;
-	indexstate->iss_RuntimeKeys = NULL;
-	indexstate->iss_NumRuntimeKeys = 0;
+    currentRelation = ExecOpenScanRelation(estate, node->scan.scanrelid, eflags);
```
Indent. use tab to replace spaces.

gpopt commented on src/backend/commands/explain.c:
```diff
@@ -3818,6 +3893,15 @@ ExplainTargetRel(Plan *plan, Index rti, ExplainState *es)
+			{
+				dynamicScanId = DynamicScan_GetDynamicScanIdPrintable(plan);
+			}
+#endif
```
We should remove this.

gpopt commented on src/backend/executor/nodeIndexscan.c:
```diff
@@ -902,190 +902,199 @@ ExecIndexRestrPos(IndexScanState *node)
+                              Relation currentRelation, Oid indexid)
+{
+    IndexScanState *indexstate;
+    LOCKMODE	lockmode;
```
I feels like my CLion IDE replaced tabs with spaces automatically when I copied this chunk of code. :(

gpopt commented on src/backend/executor/nodeIndexscan.c:
```diff
@@ -902,190 +902,199 @@ ExecIndexRestrPos(IndexScanState *node)
+    /*
+     * all done.
+     */
+    return indexstate;
```
This whole function `ExecInitIndexScanForPartition` needs reformat.

chrishajas commented on src/backend/gporca/libgpopt/src/translate/CTranslatorExprToDXL.cpp:
```diff
@@ -1630,6 +1545,98 @@ CTranslatorExprToDXL::PdxlnDynamicIndexScan(
+#endif
+
+	return pdxlnDIS;
+	/*
```
delete

kainwen commented:
@chrishajas Thanks for the PR.

Just a reminder of the issue https://github.com/greenplum-db/gpdb/issues/12533

Have you already considered that?

thedanhoffman commented:
> @chrishajas Thanks for the PR.
> 
> Just a reminder of the issue #12533
> 
> Have you already considered that?

This PR doesn't appear to address this. I'll circle back here and try my hand at a patch in a few days if it hasn't been addressed by that time.

chrishajas commented:
> @chrishajas Thanks for the PR.
> 
> Just a reminder of the issue #12533
> 
> Have you already considered that?

Thanks for linking this! I tried to repro the OOM or any increasing memory on this branch and wasn't able to. I'll play around with it a bit more to see if the issue exists.

thedanhoffman commented:
> 

The repro in the PR causes the OOM on the head of this branch. Are you generating a different plan?
```
                                              QUERY PLAN
-------------------------------------------------------------------------------------------------------
 Aggregate  (cost=0.00..782.59 rows=1 width=8)
   ->  Gather Motion 3:1  (slice1; segments: 3)  (cost=0.00..782.59 rows=1 width=4)
         ->  Nested Loop  (cost=0.00..782.59 rows=1 width=4)
               Join Filter: true
               ->  Broadcast Motion 3:3  (slice2; segments: 3)  (cost=0.00..391.30 rows=1 width=8)
                     ->  Dynamic Bitmap Heap Scan on t2  (cost=0.00..391.30 rows=1 width=8)
                           Number of partitions to scan: 100
                           Recheck Cond: (b > 50)
                           Filter: (b > 50)
                           ->  Dynamic Bitmap Index Scan on t2_b_idx  (cost=0.00..0.00 rows=0 width=0)
                                 Index Cond: (b > 50)
               ->  Dynamic Bitmap Heap Scan on t1  (cost=0.00..391.30 rows=1 width=4)
                     Number of partitions to scan: 100
                     Recheck Cond: (b > 50)
                     Filter: ((b > 50) AND (c = t2.c))
                     ->  Dynamic Bitmap Index Scan on t1_b_idx  (cost=0.00..0.00 rows=0 width=0)
                           Index Cond: (b > 50)
 Optimizer: Pivotal Optimizer (GPORCA)
(18 rows)
```

SQL
```
drop table if exists t1;
drop table if exists t2;

create table t1 (a int, b int, c text, d int)
distributed randomly
partition by range(a)
(
           start (1) end (100) every (1),
           default partition extra
);

create table t2 (a int, b int, c text, d int)
distributed randomly
partition by range(a)
(
           start (1) end (100) every (1),
           default partition extra
);

insert into t1 select  i%15, i, 'asdasdasdasasasdsdad'||i::text, i from generate_series(1, 100000)i;
insert into t2 select * from t1;

create index ON t1 USING bitmap (b);
create index ON t2 USING bitmap (b);

explain select count(distinct a.d) from t1 a, t2 b where a.c = b.c and a.b > 50 and b.b > 50;

\timing
select count(distinct a.d) from t1 a, t2 b where a.c = b.c and a.b > 50 and b.b > 50;
```

thedanhoffman commented:
> > @chrishajas Thanks for the PR.
> > Just a reminder of the issue #12533
> > Have you already considered that?
> 
> Thanks for linking this! I tried to repro the OOM or any increasing memory on this branch and wasn't able to. I'll play around with it a bit more to see if the issue exists.

The repro in the PR causes the OOM on the head of this branch. Are you generating a different plan?
```
                                              QUERY PLAN
-------------------------------------------------------------------------------------------------------
 Aggregate  (cost=0.00..782.59 rows=1 width=8)
   ->  Gather Motion 3:1  (slice1; segments: 3)  (cost=0.00..782.59 rows=1 width=4)
         ->  Nested Loop  (cost=0.00..782.59 rows=1 width=4)
               Join Filter: true
               ->  Broadcast Motion 3:3  (slice2; segments: 3)  (cost=0.00..391.30 rows=1 width=8)
                     ->  Dynamic Bitmap Heap Scan on t2  (cost=0.00..391.30 rows=1 width=8)
                           Number of partitions to scan: 100
                           Recheck Cond: (b > 50)
                           Filter: (b > 50)
                           ->  Dynamic Bitmap Index Scan on t2_b_idx  (cost=0.00..0.00 rows=0 width=0)
                                 Index Cond: (b > 50)
               ->  Dynamic Bitmap Heap Scan on t1  (cost=0.00..391.30 rows=1 width=4)
                     Number of partitions to scan: 100
                     Recheck Cond: (b > 50)
                     Filter: ((b > 50) AND (c = t2.c))
                     ->  Dynamic Bitmap Index Scan on t1_b_idx  (cost=0.00..0.00 rows=0 width=0)
                           Index Cond: (b > 50)
 Optimizer: Pivotal Optimizer (GPORCA)
(18 rows)
```

SQL
```
drop table if exists t1;
drop table if exists t2;

create table t1 (a int, b int, c text, d int)
distributed randomly
partition by range(a)
(
           start (1) end (100) every (1),
           default partition extra
);

create table t2 (a int, b int, c text, d int)
distributed randomly
partition by range(a)
(
           start (1) end (100) every (1),
           default partition extra
);

insert into t1 select  i%15, i, 'asdasdasdasasasdsdad'||i::text, i from generate_series(1, 100000)i;
insert into t2 select * from t1;

create index ON t1 USING bitmap (b);
create index ON t2 USING bitmap (b);

explain select count(distinct a.d) from t1 a, t2 b where a.c = b.c and a.b > 50 and b.b > 50;

\timing
select count(distinct a.d) from t1 a, t2 b where a.c = b.c and a.b > 50 and b.b > 50;
```

chrishajas commented:
@thedanhoffman Thanks! Now I see it (I had to reduce the data, as before it would just take a long time before even starting to accumulate memory). I see the issue for all dynamic scans.

gpopt commented on src/backend/executor/nodeDynamicSeqscan.c:
```diff
@@ -160,6 +96,8 @@ ExecInitDynamicSeqScan(DynamicSeqScan *node, EState *estate, int eflags)
 									 ALLOCSET_DEFAULT_MAXSIZE);
 
+	
+
```
remove this

dgkimura commented on src/backend/executor/nodeDynamicBitmapHeapscan.c:
```diff
@@ -161,18 +192,15 @@ initNextTableToScan(DynamicBitmapHeapScanState *node)
-		scanState->ps.targetlist = (List *) ExecInitExpr((Expr *) scanState->ps.plan->targetlist,
-														 (PlanState *) scanState);
+		scanState->ps.qual =
+				ExecInitQual(scanState->ps.plan->qual, (PlanState *) scanState);
```
Comment above this block regarding `targetlist` now seems inaccurate.  Also, if it's no longer true, then where/when is the `targetlist` initialized?

chrishajas replied:
Fixed. The plan state no longer has a targetlist as of b8d7f053c5c2bf2a7e8734fe3327f6a8bc711755, so initializing it here isn't needed.

dgkimura reviewed (commented):
Awesome work!

I'm still in the middle of reviewing and digesting this PR.  I'm at commit https://github.com/greenplum-db/gpdb/pull/13551/commits/bf0ded346564d1f96a05a74cb5777100e06fcc35.   Just posting the comments that I have so far.

dgkimura commented on src/backend/commands/explain.c:
```diff
@@ -1855,27 +1853,32 @@ ExplainNode(PlanState *planstate, List *ancestors,
-					indexoid = parentOid;
-					parentOid = rel_partition_get_root(indexoid);
-				}
-
```
Do we no longer need the partition root because... we no longer build logical index info from the root partition?

chrishajas replied:
It's more that we aren't currently supporting multi-level partitioning. We'd need to use a different function, so I think we can add this logic back once we support multi-level.

dgkimura replied:
Ah, I think that makes sense. Thanks for that insight.

dgkimura commented on src/backend/executor/nodeDynamicSeqscan.c:
```diff
@@ -8,13 +8,7 @@
- * partOids field in the DynamicSeqScan plan node that holds the selected
- * partitions; but none of the code in this file has been fixed to
- * actually work that way. Same with all the other Dynamic*Scan nodes.
+ * This has a smaller plan size than using an append with many partitions
```
To make sure I follow, the smaller plan size after this commit is largely attributed to only needing a single dynamic scan node to represent column information across the entire partitioned table?  Whereas before this change we used a scan node for each partition, each of which holds essentially duplicate column information of the table?  If this is true though, how are dropped columns accounted for in both models?

chrishajas replied:
Added a bit more to the comment. But yes, it's due to only having a single dynamic node in the plan, and not having to populate a target list for every individual partition. For many partitions with many columns, this made the plan size very large (100+ MB in some cases). We handle dropped/remapped columns during execution now, see `convert_tuples_by_name_map_if_req`

dgkimura reviewed (commented):
Looks good to me.  No blocking concerns I could think of.  I left a few questions/comments.  Would like to know if/how we handle dropped columns.  Looking forward to getting this into master and iterating.

chrishajas commented on src/include/nodes/execnodes.h:
```diff
@@ -1607,31 +1611,22 @@ typedef struct DynamicIndexScanState
 	/* The partition oid for which the current varnos are mapped */
 	Oid columnLayoutOid;
+
+	struct PartitionPruneState *as_prune_state;
+	Bitmapset  *as_valid_subplans;
```
I'm not sure actually, append scan would be my best guess

chrishajas commented on src/backend/executor/nodeDynamicSeqscan.c:
```diff
@@ -150,6 +79,13 @@ ExecInitDynamicSeqScan(DynamicSeqScan *node, EState *estate, int eflags)
 
 	state->scanrelid = node->seqscan.scanrelid;
 
+	state->as_valid_subplans =
+		bms_add_range(NULL, 0, list_length(node->partOids) - 1);
+	state->as_prune_state = NULL;
+
+	if (node->join_prune_paramids)
+		state->as_valid_subplans = NULL;
```
Nice catch! We initially copied it from the Append node, which scans these subplans, however we're doing things slightly differently so we don't need to populate it here.

chrishajas commented on src/backend/executor/nodeIndexscan.c:
```diff
@@ -902,190 +902,199 @@ ExecIndexRestrPos(IndexScanState *node)
 IndexScanState *
 ExecInitIndexScan(IndexScan *node, EState *estate, int eflags)
 {
```
Fixed up the tabs/spaces, so hopefully there will be less of a diff

chrishajas commented on src/include/nodes/execnodes.h:
```diff
@@ -1970,6 +2098,48 @@ typedef struct ForeignScanState
+
+	int			nOids;
+	Oid		   *partOids;
+	int			whichPart;
```
Yep, adding a comment for these

chrishajas commented on src/backend/executor/nodeBitmapHeapscan.c:
```diff
@@ -752,8 +752,37 @@ ExecEndBitmapHeapScan(BitmapHeapScanState *node)
+	 *
+	 * We do this last because the child nodes will open indexscans on our
+	 * relation's indexes, and we want to be sure we have acquired a lock on
+	 * the relation first.
```
I'm not certain on this, but I think it's saying we need to lock the relation before we open indexscans on it, possibly to ensure data consistency? This was brought back from 6X, so I'm not entirely sure though.

chrishajas commented on src/backend/executor/nodeDynamicSeqscan.c:
```diff
@@ -337,5 +304,7 @@ ExecReScanDynamicSeqScan(DynamicSeqScanState *node)
 
-	/* Force reloading the partition hash table */
+	// reset partition internal state
+	node->whichPart = -1;
```
Yes I believe so. Adding that logic.

chrishajas commented on src/backend/gpopt/translate/CTranslatorDXLToPlStmt.cpp:
```diff
@@ -3870,9 +3870,36 @@ CTranslatorDXLToPlStmt::TranslateDXLDynTblScan(
+
+	IMdIdArray *parts = dyn_tbl_scan_dxlop->GetParts();
+
+	List *oids_list = NIL;
+
+	for (ULONG ul = 0; ul < parts->Size(); ul++)
+	{
+		Oid part = CMDIdGPDB::CastMdid((*parts)[ul])->Oid();
+		oids_list = gpdb::LAppendOid(oids_list, part);
+	}
+
+	dyn_seq_scan->partOids = oids_list;
```
I went through the path of changing partOids to a "vector" instead. If we do this we'd need to also populate the number of partOids, which otherwise is stored in the list, as that's used in the copy/outfuncs and a couple other places in code. I can add this if you think it's more readable.

chrishajas commented on src/backend/gporca/libnaucrates/src/parser/CParseHandlerIndexScan.cpp:
```diff
@@ -163,6 +174,10 @@ CParseHandlerIndexScan::StartElementHelper(
 	this->Append(proj_list_parse_handler);
 	this->Append(filter_parse_handler);
 	this->Append(index_condition_list_parse_handler);
+	if (token_type == EdxltokenPhysicalDynamicIndexScan)
+	{
+		this->Append(partition_mdids_parse_handler);
+	}
```
Yeah, that's the reasoning. If we added that to index scans, it would get rid of the token checks and we'd be able to hardcode the numbers again. On the other hand, It would be an always-empty field for index scans and I believe we'd need to update mdps. I don't have a strong opinion either way in this case.

dgkimura replied:
Iff it's trivial to automate the MDP fix-up and the cost of the extra field in the MDP is insignificant then I'd vote to change.  But if you think either of those isn't true then I'm okay leaving as-is.

chrishajas replied:
Done!

chrishajas commented on src/backend/executor/nodeDynamicSeqscan.c:
```diff
@@ -264,6 +203,34 @@ ExecDynamicSeqScan(PlanState *pstate)
+									 list_length(plan->partOids),
+									 plan->join_prune_paramids);
+
+		int i = 0;
+		int partOidIdx = -1;
+		List *newPartOids = NIL;
+		ListCell *lc;
+		for(i = 0; i < bms_num_members(node->as_valid_subplans); i++)
+		{
+			partOidIdx = bms_next_member(node->as_valid_subplans, partOidIdx);
+			newPartOids = lappend_oid(newPartOids, node->partOids[partOidIdx]);
+		}
+
+		node->partOids = palloc(sizeof(Oid) * list_length(newPartOids));
+		foreach_with_count(lc, newPartOids, i)
+			node->partOids[i] = lfirst_oid(lc);
```
Done, much better thanks!

chrishajas replied:
Actually, we do need the separate list. `node->as_valid_subplans` is the indexes of the partitions that should be selected. `node->partOids` is the static list of partitions to scan. We overwrite `node->partOids` with the dynamically selected indexes, but first we need to get the oids from `node->partOids`.

gpopt commented on src/backend/executor/nodeDynamicBitmapHeapscan.c:
```diff
@@ -0,0 +1,338 @@
+	 */
+	attMap = convert_tuples_by_name_map_if_req(partTupDesc, lastTupDesc, "unused msg");
+	table_close(lastScannedRel, AccessShareLock);
+//	attMap = IndexScan_GetColumnMapping(*pid, node->lastRelOid);
```
Remove this line

gpopt commented on src/backend/executor/nodeDynamicIndexscan.c:
```diff
@@ -0,0 +1,343 @@
+ExecInitDynamicIndexScan(DynamicIndexScan *node, EState *estate, int eflags)
+{
+	DynamicIndexScanState *dynamicIndexScanState;
+    ListCell *lc;
```
The indentation in this function has not been fixed.

chrishajas commented:
@dgkimura @gpopt Thanks for the reviews! I've addressed the feedback or replied to it all (I think!). If there's anything else let me know. For now I have it as a separate commit, but I'll squash the areas into the appropriate part before it's merged (along with rebasing to resolve some expected file changes).

chrishajas commented:
> To make sure I follow, the smaller plan size after this commit is largely attributed to only needing a single dynamic scan node to represent column information across the entire partitioned table? Whereas before this change we used a scan node for each partition, each of which holds essentially duplicate column information of the table? If this is true though, how are dropped columns accounted for in both models?

@dgkimura Correct,  we're no longer storing the individual columns of each partition. Dropped/remapped column information therefore also isn't stored in the plan. Instead, they're determined during execution.  
```
/* If attribute remapping is not necessary, then do not change the varattno */
	if (attMap)
	{
		change_varattnos_of_a_varno((Node*)scanState->ps.plan->qual, attMap, node->scanrelid);
		change_varattnos_of_a_varno((Node*)scanState->ps.plan->targetlist, attMap, node->scanrelid);

```

dgkimura commented on src/backend/executor/nodeIndexscan.c:
```diff
@@ -921,20 +935,15 @@ ExecInitIndexScan(IndexScan *node, EState *estate, int eflags)
 	ExecInitScanTupleSlot(estate, &indexstate->ss,
-						  RelationGetDescr(currentRelation),
-						  table_slot_callbacks(currentRelation));
+		                  RelationGetDescr(currentRelation),
+		                  table_slot_callbacks(currentRelation));
```
Looks like there's still a handful of spaces versus tabs changes in this file

dgkimura reviewed (approved):
Awesome work!  I think this is a great starting point to iterate on.
